src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py

```diff
@@ -0,0 +1,322 @@
+            if latest_task_status in [JobStatus.RUNNING, JobStatus.PENDING]:
+                self._cancel_job(job_id)
+
+            return True
+
+        if latest_task['source_data_dir'] != self.source_data_dir:
+            raise PreconditionError(
+                f'There is a previous run of the same tool and load version with a different source data directory.')
+
+        if latest_task_status in [JobStatus.FAILED, JobStatus.CANCELLED]:
+            print(f'The tool and load version have been run before, '
+                  f'and the most recent status is {str(latest_task_status)}.'
+                  f'Resuming progress from the previous run.')
+        elif latest_task_status in [JobStatus.COMPLETED, JobStatus.RUNNING, JobStatus.PENDING]:
+            raise PreconditionError(
+                f'There is a previous run of the same tool and load version that is {str(latest_task_status)}.')
+        else:
+            raise ValueError(f'Unexpected job status: {latest_task_status}')
+
+    def __init__(self, kbase_collection, load_ver, tool, source_data_dir, root_dir=loader_common_names.ROOT_DIR):
+        """
+        Initialize the task manager.
+
+        :param kbase_collection: KBase collection identifier name (e.g. GTDB).
+        :param load_ver: collection load version (e.g. r207.kbase.1).
+        :param tool: tool name (e.g. gtdb_tk, checkm2, etc.)
+        :param source_data_dir: source data directory.
+        :param root_dir: root directory of the collection project.
+                         Default is the ROOT_DIR defined in src/loaders/common/loader_common_names.py
+
+        """
+
+        self.kbase_collection = kbase_collection
+        self.load_ver = load_ver
+        self.tool = tool
+        self.source_data_dir = source_data_dir
+        self.root_dir = root_dir
+
+        # job directory is named as <kbase_collection>_<load_ver>_<tool>
+        self.job_dir = os.path.join(self.root_dir, TASKFARMER_JOB_DIR,
+                                    f'{self.kbase_collection}_{self.load_ver}_{self.tool}')
+
+    def submit_job(self, restart_on_demand=False):
+        """
+        Submit the job to slurm
+
+        Follow the steps in https://docs.nersc.gov/jobs/workflow/taskfarmer/#taskfarmer and generate all the necessary
+        files for the job submission (e.g. wrapper script, task list and batch script, etc.) before calling this function.
+
+        :param restart_on_demand: if True, killed any running/pending jobs and restart running the tool again.
+        """
+
+        # check if all the required files exist
+        required_files = [tf_common.WRAPPER_FILE, tf_common.TASK_FILE, tf_common.BATCH_SCRIPT]
+        for filename in required_files:
+            if not os.path.isfile(os.path.join(self.job_dir, filename)):
+                raise ValueError(f"{filename} does not exist in {self.job_dir}")
+
+        self._check_preconditions(restart_on_demand)
+
+        os.chdir(self.job_dir)
```

Comment:
Why is this necessary? The job dir is provided to nersc running function

Reply:
this is to make sure slurm logs will be created in side the job dir. Otherwise, it will be created where you run task_generator script.

Reply:
Why not just pass the job dir in the `cwd` argument to subprocess? https://docs.python.org/3/library/subprocess.html#subprocess.run

Reply:
I guess they are the same thing.

Reply:
Except that if you do it in the function it's less likely to cause bugs, because ppl calling the function may not realize they need to chdir themselves. Makes the function self contained and easier to use

Reply:
👍

Reply:
Why `shell` vs `cwd`? `shell` has security issues and so should be avoided if possible
https://docs.python.org/3/library/subprocess.html#security-considerations

Reply:
Also the current code still required the caller of the run_nersc_command option to run a cd, which is what I wanted to avoid

Reply:
IOW
```
p = subprocess.Popen(command, stdout=std_out, stderr=std_err, text=True, cwd=job_dir)```

Reply:
avoid 'cd' completely?

Reply:
> any strong reasons, we want to use subprocess? It will create a log file if I use the helper run_command function. Is that worth it for a cd command?

Sorry I missed this comment earlier. I don't want to call the run nersc function to do the cd, I want, effectively, the cd to happen in the run nersc function like above

Reply:
Oh, you did this in the latest commit. Lots of race conditions here

Reply:
:+1: